contrib/nitiwiki/src/wiki_edit.nit

```diff
@@ -0,0 +1,192 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Web server to server generated files and modify the wiki from a web form
+module wiki_edit
+
+import nitcorn
+import markdown
+import md5
+
+intrude import wiki_html
+
+# Page for editing markdown source
+class WikiEditForm
+	super WikiArticle
+
+	# Part of the title before the name of the page
+	var title_prefix: String
+
+	# Markdown content, for previews
+	redef var md
+
+	# Custom HTML code, for forms and links
+	var html: String
+
+	init do content = (md or else "").md_to_html.to_s + html
+
+	redef fun dir_href do return "edit" / href
+
+	redef fun tpl_article
+	do
+		var s = super
+		s.title = title_prefix + title
+		return s
+	end
+
+	# Fill and return a new `HttpResponse` with this page content
+	fun to_http_response: HttpResponse
+	do
+		var resp = new HttpResponse(200)
+		resp.body = tpl_page.write_to_string
+		return resp
+	end
+end
+
+# Action to serve edit forms, show previews and apply changes
+class EditAction
+	super Action
+
+	# Full public URL for the root of this wiki
+	var root_url: String
+
+	# Path to the wiki config
+	var config_file_path: String
+
+	# Configuration of the Wiki, loaded once
+	var wiki_config = new WikiConfig(config_file_path) is lazy
+
+	# Path to the root of the wiki
+	private var wiki_root: String = config_file_path.dirname is lazy
+
+	# Path to the source files
+	private var source_dir: String = (wiki_root / wiki_config.source_dir).simplify_path + "/" is lazy
+
+	# List of acceptable password to apply modifications
+	#
+	# If `null`, no password checks are applied and all modifications are accepted.
+	var passwords: nullable Collection[String]
+
+	# Reload the wiki instance with the latest changes
+	fun wiki: Nitiwiki
+	do
+		var wiki = new Nitiwiki(wiki_config)
+		wiki.parse
+		return wiki
+	end
+
+	redef fun answer(http_request, turi)
+	do
+		var action = http_request.string_arg("action")
+		var markdown = http_request.post_args.get_or_default("content", "")
+
+		var file_path = turi.strip_leading_slash
+		file_path = wiki_root / file_path
+
+		if not file_path.simplify_path.has_prefix(source_dir) then
```

Comment:
Seems to be unsafe.

The check depends on source dir. if source dir == "." it will even match something like "../../../etc/passwd"